distributed/worker_state_machine.py

```diff
@@ -1217,54 +1211,78 @@ def _purge_state(self, ts: TaskState) -> None:
         self.executing.discard(ts)
         self.in_flight_tasks.discard(ts)
 
+    def _reorder_data_needed(self) -> None:
+        """Helper of `_ensure_communicating`.
+
+        Reorder peer workers in data_needed by:
+
+        1. local -> remote -> busy or in flight
+        2. if not busy and tied, by highest-priority task available
+        3. if not busy and tied, by number of tasks available to be fetched
+        4. if not busy and tied, by a random element
+
+        Remove peer workers with no tasks.
+        """
+        host = get_address_host(self.address)
+        to_sort = []
+        busy = {}
+        for worker, tasks in self.data_needed.items():
+            if not tasks:
+                continue
+            if worker in self.in_flight_workers or worker in self.busy_workers:
+                busy[worker] = tasks
+                continue
+
+            to_sort.append(
+                (
+                    get_address_host(worker) != host,  # False < True
+                    tasks.peek().priority,
+                    -len(tasks),
+                    random.random(),
+                    worker,
+                    tasks,
+                )
+            )
+
+        self.data_needed.clear()
+        self.data_needed.update(
+            (worker, tasks) for _, _, _, _, worker, tasks in sorted(to_sort)
+        )
```

Comment:
Note that, if your run `_ensure_communicating` twice in a row, on the second round `to_sort` will be empty